### PR TITLE
coprocessor/codec: remove read trait from datum and table in decode f…

### PR DESCRIPTION
**Pull request by AndreMouche**
hi, this PR  remove the read trait from datum and table in decode function.
@BusyJay @breeswish @hicqu PTAL

### CODE REVIEW

File: src/coprocessor/codec/datum.rs [hunk 1]
```diff
@@ -759,45 +759,44 @@ impl From<Json> for Datum {
...
 
-pub trait DatumDecoder: DecimalDecoder + JsonDecoder {
-    /// `decode_datum` decodes on a datum from a byte slice generated by tidb.
-    fn decode_datum(&mut self) -> Result<Datum> {
-        let flag = self.read_u8()?;
+/// `decode_datum` decodes on a datum from a byte slice generated by tidb.
+pub fn decode_datum(data: &mut BytesSlice) -> Result<Datum> {
+    if !data.is_empty() {
+        let flag = data[0];
+        *data = unsafe { data.get_unchecked(1..) };
```
Comment (breeswish) [inline src/coprocessor/codec/datum.rs]: What is this used for? Or, do we have ways without unsafe?

Comment (BusyJay) [reply src/coprocessor/codec/datum.rs]: Use `&data[1..]` is the same here. Because `data.is_empty()` has provide the hint to compiler.

**Comment by breeswish** [comment]
LGTM

**Comment by AndreMouche** [comment]
friendly ping @BusyJay @hicqu

**Comment by hicqu** [comment]
LGTM.

**Pull request merged**